### ISSUE TITLE
Segfault or silently wrong result on differentiation of GP loss function

### ISSUE THREAD

**Issue by simsurace** [EnzymeAD/Enzyme.jl#2355]
This came up while investigating [this topic on Discourse](https://discourse.julialang.org/t/large-memory-consumption-when-using-mooncake-via-differentiationinterface-for-gaussian-process-optimisation/127289/18). The following script iterates over a few backends and prints gradients. Depending on the size of `N` (and luck), this will either give different results for the Enzyme backends vs. FiniteDiff or Zygote or even segfault. When it segfaulted, I got the following error:

```
[74750] signal (11.2): Segmentation fault: 11
in expression starting at REPL[19]:1
.Ldgemm_tcopy_L8_M8_80 at ~/.julia/juliaup/julia-1.10.9+0.aarch64.apple.darwin14/lib/julia/libopenblas64_.dylib (unknown line)
Allocations: 87827445 (Pool: 87719349; Big: 108096); GC: 101
[1]    74750 segmentation fault  julia +lts --project=. -t4
```

```julia
using AbstractGPs, DifferentiationInterface, Optim, StatsFuns, Random

const DI = DifferentiationInterface

import Enzyme
import FiniteDiff
import Zygote

rng = MersenneTwister(1)

N = 1000
x = randn(rng, N)
y = randn(rng, N)

function loss_function(x, y)
    function negativelogmarginallikelihood(params)
        kernel =
            softplus(params[1]) * (Matern52Kernel() ∘ ScaleTransform(softplus(params[2])))
        noise_var = 0.1
        f = GP(kernel)
        fx = f(x, noise_var)
        return -logpdf(fx, y)
    end
    return negativelogmarginallikelihood
end

θ0 = randn(rng, 2)

losses = [
    loss_function(x[1:4], y[1:4]),
    loss_function(x, y)
]

backends = [
    AutoFiniteDiff(),
    AutoZygote(),
    AutoEnzyme(
        mode=Enzyme.Reverse,
        function_annotation=Enzyme.Duplicated,
    ),
    AutoEnzyme(
        mode=Enzyme.set_runtime_activity(Enzyme.Reverse),
        function_annotation=Enzyme.Const,
    ),
]

for loss in losses
    for backend in backends
        grad = DI.gradient(loss, backend, θ0)
        @info "Gradient" backend grad
    end
end

```

```julia
julia> versioninfo()
Julia Version 1.10.9
Commit 5595d20a287 (2025-03-10 12:51 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: macOS (arm64-apple-darwin24.0.0)
  CPU: 12 × Apple M3 Pro
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, apple-m1)
Threads: 4 default, 0 interactive, 2 GC (on 6 virtual cores)
Environment:
  JULIA_EDITOR = code
```

**Comment by wsmoses** [comment]
Thanks for the report!

This input code is rather complex, can you try to simplify it (ideally being as close to pure julia as possible, using as few as possible external packages as possible), and directly calling Enzyme.autodiff?

This will allow us to find and fix the error fastest

**Comment by simsurace** [comment]
This is essentially a Gaussian likelihood with a parameterized covariance matrix. There's a cholesky factorization and some linalg. I'll try to minimize it.

**Comment by simsurace** [comment]
This is a MRE for the wrong gradient. No segfaults with this one. The real code above makes a call to a function from Distances.jl with `@inbounds`, so that could be the problem.
```julia
using LinearAlgebra
import DifferentiationInterface as DI
import Enzyme
import FiniteDiff

function loss_function(x)
    function testfunction(params)
        xs = params[1] * x
        C_mat = zeros(length(xs), length(xs))
        mul!(C_mat, xs, xs')
        return sum(C_mat)
    end
    return testfunction
end

θ0 = randn(1)
x = randn(10_000)
loss = loss_function(x)

backends = [
    DI.AutoFiniteDiff(),
    DI.AutoEnzyme(
        mode=Enzyme.Reverse,
        function_annotation=Enzyme.Duplicated,
    ),
    DI.AutoEnzyme(
        mode=Enzyme.set_runtime_activity(Enzyme.Reverse),
        function_annotation=Enzyme.Const,
    ),
]

for backend in backends
    grad = DI.gradient(loss, backend, θ0)
    @info "Gradient" backend grad
end
```

**Comment by wsmoses** [comment]
@simsurace just to confirm does this still err (a lot of relevant fixes have gone in since)

**Comment by simsurace** [comment]
I just checked the last example even still produces very different results compared to FiniteDiff:
```julia
julia> enzyme_grad = let dθ = Enzyme.make_zero(θ0)
           Enzyme.autodiff(
               Enzyme.Reverse,
               Enzyme.Duplicated(loss, Enzyme.make_zero(loss)),
               Enzyme.Duplicated(θ0, dθ),
           )
           dθ
       end
 ** On entry to DSYMM  parameter number  9 had an illegal value
1-element Vector{Float64}:
 -15147.222392413192

julia> finitediff_grad = FiniteDiff.finite_difference_gradient(loss, θ0)
1-element Vector{Float64}:
 -2455.863678399372

julia> enzyme_grad ≈ finitediff_grad # false
false

(jl_swi6WU) pkg> st
Status `/private/var/folders/31/r_tx2_5d4hn30gf8qbb0hjjc0000gp/T/jl_swi6WU/Project.toml`
  [7da242da] Enzyme v0.13.76
  [6a86dc24] FiniteDiff v2.28.1
  [37e2e46d] LinearAlgebra v1.11.0

julia> versioninfo()
Julia Version 1.11.6
Commit 9615af0f269 (2025-07-09 12:58 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: macOS (arm64-apple-darwin24.0.0)
  CPU: 12 × Apple M3 Pro
  WORD_SIZE: 64
  LLVM: libLLVM-16.0.6 (ORCJIT, apple-m2)
Threads: 1 default, 0 interactive, 1 GC (on 6 virtual cores)
```

**Comment by simsurace** [comment]
Wouldn't this have to be covered by unit tests? It's just a `mul!` and a `sum`.

**Comment by simsurace** [comment]
Where would one look for the rule that gets called for this call (presumably BLAS)?

**Comment by wsmoses** [comment]
should be fixed by https://github.com/EnzymeAD/Enzyme.jl/pull/2937

**Comment by simsurace** [comment]
I just checked, but sadly the first example still segfaults and the last example still gives the wrong gradient.

**Comment by wsmoses** [comment]
Can you run up and then show package version? I did explicitly test the last code (and understood why it mattered), so either the setup is wrong or I failed to merge and release or something

**Comment by simsurace** [comment]
On main:
```julia
julia> using TestEnv; TestEnv.activate()
"/var/folders/31/r_tx2_5d4hn30gf8qbb0hjjc0000gp/T/jl_aw8IuE/Project.toml"

julia> using LinearAlgebra

julia> import Enzyme

julia> import FiniteDifferences

julia> function loss_function(x)
           function testfunction(params)
               xs = params[1] * x
               C_mat = zeros(length(xs), length(xs))
               mul!(C_mat, xs, xs')
               return sum(C_mat)
           end
           return testfunction
       end
loss_function (generic function with 1 method)

julia> θ0 = randn(1)
1-element Vector{Float64}:
 0.6906165421011959

julia> x = randn(10_000);

julia> loss = loss_function(x)
(::var"#testfunction#11"{Vector{Float64}}) (generic function with 1 method)

julia> enzyme_grad = let dθ = Enzyme.make_zero(θ0)
           Enzyme.autodiff(
               Enzyme.Reverse,
               Enzyme.Duplicated(loss, Enzyme.make_zero(loss)),
               Enzyme.Duplicated(θ0, dθ),
           )
           dθ
       end
 ** On entry to DSYMM  parameter number  9 had an illegal value
1-element Vector{Float64}:
 6819.984450443334

julia> finitediff_grad = FiniteDifferences.grad(FiniteDifferences.central_fdm(5, 1), loss, θ0) |> only
1-element Vector{Float64}:
 13979.518331745674

julia> enzyme_grad ≈ finitediff_grad # false
false

julia> versioninfo()
Julia Version 1.11.7
Commit f2b3dbda30a (2025-09-08 12:10 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: macOS (arm64-apple-darwin24.0.0)
  CPU: 12 × Apple M3 Pro
  WORD_SIZE: 64
  LLVM: libLLVM-16.0.6 (ORCJIT, apple-m2)
Threads: 4 default, 0 interactive, 2 GC (on 6 virtual cores)
Environment:
  JULIA_EDITOR = code
  JULIA_VSCODE_REPL = 1
  JULIA_NUM_THREADS = 4

(jl_aw8IuE) pkg> st
Status `/private/var/folders/31/r_tx2_5d4hn30gf8qbb0hjjc0000gp/T/jl_aw8IuE/Project.toml`
  [ab4f0b2a] BFloat16s v0.6.1
  [082447d4] ChainRules v1.72.6
  [d360d2e6] ChainRulesCore v1.26.0
  [e30172f5] Documenter v1.16.1
  [7da242da] Enzyme v0.13.125 `~/.julia/dev/Enzyme`
  [12d8515a] EnzymeTestUtils v0.2.6 `~/.julia/dev/Enzyme/lib/EnzymeTestUtils`
  [26cc04aa] FiniteDifferences v0.12.33
  [61eb1bfa] GPUCompiler v1.8.2
  [842dd82b] InlineStrings v1.4.5
  [27aeb0d3] JLArrays v0.3.1
  [929cbde3] LLVM v9.4.4
  [2ab3a3ac] LogExpFunctions v0.3.29
  [d3525ed8] ParallelTestRunner v2.1.0
  [276daf66] SpecialFunctions v2.6.1
  [90137ffa] StaticArrays v1.9.16
  [10745b16] Statistics v1.11.1
  [7cc45869] Enzyme_jll v0.0.245+0
⌅ [86de99a1] LLVM_jll v16.0.6+5
  [37e2e46d] LinearAlgebra v1.11.0
  [44cfe95a] Pkg v1.11.0
  [9a3f8284] Random v1.11.0
  [2f01184e] SparseArrays v1.11.0
  [fa267f1f] TOML v1.0.3
  [8dfed614] Test v1.11.0
Info Packages marked with ⌅ have new versions available but compatibility constraints restrict them from upgrading. To see why use `status --outdated`
```

**Comment by wsmoses** [comment]
can you free Enzyme, and not use a dev'd version?

**Comment by wsmoses** [comment]
but regardless will try to repro locally shortly

**Comment by simsurace** [comment]
Hmm, not sure what you mean by "free". I'm in the Enzyme repo, current main checked out (15fb5dff194e9bf75202f7e62491286ebb472402).

**Comment by wsmoses** [comment]
lets see if https://github.com/EnzymeAD/Enzyme.jl/pull/3015 resolves

**Comment by wsmoses** [comment]
@simsurace can you retry and see if resolved [essentially the same issue was just resolved for a different case]

**Comment by wsmoses** [comment]
I'm not seeing this reproduce locally @simsurace  so I'm going to close. Please reopen if it persists
